torch/csrc/generic/StorageSharing.cpp

```diff
@@ -212,27 +212,49 @@ static PyObject * THPStorage_(newSharedFd)(PyObject *_unused, PyObject *args)
 
 #else // THC_GENERIC_FILE
 
-static PyObject * THPStorage_(shareCuda)(THPStorage *self)
-{
+static PyObject* THPStorage_(shareCuda)(THPStorage* self) {
   HANDLE_TH_ERRORS
-  THWStorage *storage = self->cdata;
+  THWStorage* storage = self->cdata;
+
+  if (storage->received_cuda()) {
+    AT_ERROR(
+        "Attempting to send tensor received from another process. Consider clonning.");
+  }
+
   at::DeviceGuard device_guard(storage->device());
-  THPObjectPtr tuple(PyTuple_New(4));
+  THPObjectPtr tuple(PyTuple_New(5));
   THPObjectPtr device(PyLong_FromLong(storage->device().index()));
   THPObjectPtr _handle(Py_None);
   Py_INCREF(Py_None);
   THPObjectPtr size_bytes(PyLong_FromLong(storage->numel() * sizeof(scalar_t)));
   THPObjectPtr _offset_bytes(PyLong_FromLong(0));
+  THPObjectPtr _ref_counter(Py_None);
+  Py_INCREF(Py_None);
   if (THWStorage_(data)(LIBRARY_STATE storage)) {
     size_t base_size;
-    void *base_ptr = c10::cuda::CUDACachingAllocator::getBaseAllocation(THWStorage_(data)(LIBRARY_STATE storage), &base_size);
+    void* base_ptr = c10::cuda::CUDACachingAllocator::getBaseAllocation(
+        THWStorage_(data)(LIBRARY_STATE storage), &base_size);
     ptrdiff_t offset_bytes = (char*)storage->data<scalar_t>() - (char*)base_ptr;
 
     cudaIpcMemHandle_t handle;
     THCudaCheck(cudaIpcGetMemHandle(&handle, base_ptr));
 
-    _handle = PyBytes_FromStringAndSize((char *)&handle, CUDA_IPC_HANDLE_SIZE);
+    _handle = PyBytes_FromStringAndSize((char*)&handle, CUDA_IPC_HANDLE_SIZE);
     _offset_bytes = PyLong_FromSsize_t((Py_ssize_t)offset_bytes);
+
+    int flags = TH_ALLOCATOR_MAPPED_SHAREDMEM | TH_ALLOCATOR_MAPPED_EXCLUSIVE;
+    std::string ref_counter_handle = THPStorage_(__newHandle)();
+    at::DataPtr sptr = THRefcountedMapAllocator::makeDataPtr(
+        ref_counter_handle.c_str(), flags, sizeof(int64_t), nullptr);
+    *(int64_t*)(sptr.get()) += 1;
+
+    auto ptr = new CudaIPCSentData(std::move(sptr));
+    at::DataPtr new_data_ptr = at::DataPtr(
+        storage->data(), ptr, CudaIPCSentDataDelete, at::DeviceType::CPU);
```

Comment:
I think the DeviceType here is wrong. Even though the context pointer `ptr` is CPU, the `storage->data()` pointer is still a CUDA pointer, and the device type should reflect that.

Reply:
How about taking current device from the storage?

Reply:
SGTM